apps/passport-server/src/services/frogcryptoService.ts

```diff
@@ -0,0 +1,272 @@
+import { FrogCryptoUserFeedState } from "../database/models";
+import {
+  fetchUserFeedsState,
+  getFrogData,
+  initializeUserFeedState,
+  insertFrogData,
+  sampleFrogData,
+  updateUserFeedState
+} from "../database/queries/frogcrypto";
+import { fetchUserByCommitment } from "../database/queries/users";
+import { sqlTransaction } from "../database/sqlQuery";
+import { PCDHTTPError } from "../routing/pcdHttpError";
+import { ApplicationContext } from "../types";
+import { FROGCRYPTO_FEEDS, FrogCryptoFeedConfig } from "../util/frogcrypto";
+import { logger } from "../util/logger";
+import { RollbarService } from "./rollbarService";
+
+export class FrogcryptoService {
+  private readonly context: ApplicationContext;
+  private readonly rollbarService: RollbarService | null;
+  private readonly verificationPromiseCache: LRUCache<
+    string,
+    Promise<string | null>
+  >;
+
+  public constructor(
+    context: ApplicationContext,
+    rollbarService: RollbarService | null
+  ) {
+    this.context = context;
+    this.rollbarService = rollbarService;
+    this.verificationPromiseCache = new LRUCache<
+      string,
+      Promise<string | null>
+    >({
+      max: 1000
+    });
+  }
+
+  public async getFeeds(): Promise<FrogCryptoFeedConfig[]> {
+    return FROGCRYPTO_FEEDS;
+  }
+
+  public async getUserState(
+    req: FrogCryptoUserStateRequest
+  ): Promise<FrogCryptoUserStateResponseValue> {
+    const semaphoreId = await this.cachedVerifySignaturePCD(req.pcd);
+    if (!semaphoreId) {
+      throw new PCDHTTPError(400, "invalid PCD");
+    }
+
+    const userFeeds = await fetchUserFeedsState(
+      this.context.dbPool,
+      semaphoreId
+    );
+
+    const allFeeds = _.keyBy(await this.getFeeds(), "id");
+
+    return {
+      feeds: userFeeds.map((userFeed) =>
+        this.computeUserFeedState(userFeed, allFeeds[userFeed.feed_id])
+      )
+    };
+  }
+
+  public async reserveFrogData(
+    pcd: SerializedPCD<SemaphoreSignaturePCD>,
+    feed: FrogCryptoFeedConfig
+  ): Promise<IFrogData> {
+    const semaphoreId = await this.cachedVerifySignaturePCD(pcd);
+    if (!semaphoreId) {
+      throw new PCDHTTPError(400, "invalid PCD");
+    }
+
+    await initializeUserFeedState(this.context.dbPool, semaphoreId, feed.id);
+
+    return sqlTransaction(
+      this.context.dbPool,
+      "reserve frog",
+      async (client) => {
+        const lastFetchedAt = await updateUserFeedState(
+          client,
+          semaphoreId,
+          feed.id
+        ).catch((e) => {
+          if (e.message.includes("could not obtain lock")) {
+            throw new PCDHTTPError(
+              429,
+              "There is another frog request in flight!"
+            );
+          }
+          throw e;
+        });
+        if (!lastFetchedAt) {
+          const e = new Error("User feed state unexpectedly not found!");
+          logger(`Error encountered while serving feed:`, e);
+          this.rollbarService?.reportError(e);
```

Comment:
```suggestion
```

given you're `throw e`ing below, you don't actually need to do this `reportError` here, the error would be reported in an earlier scope.